### PR TITLE
Task 4 (Integration with Database)

### PR DESCRIPTION
**Pull request by nadktk**
_Dear reviewers, thanks for giving me more time. It was extremely hard to find some extra time during last week. So, the solution needs some clean up and improvements from me, but at least everything is done for today._

# What was done?

- [x] (1) Task 4.1 is implemented (Use AWS Console to create a database instance in RDS with PostgreSQL, create tables and fill them with script, save script in the repo)
- [x] (2) - Task 4.2 is implemented lambda links are provided and returns data (integrate lambdas with DB)
- [x] (4) - Task 4.3 is implemented lambda links are provided and products is stored in DB (Create a lambda function called createProduct to be triggered by the HTTP POST method)
- [x] (5) - Your own Frontend application is integrated with Product Service (/products API) and products from Product Service are represented on Frontend. Link to a working Frontend application is provided for cross-check reviewer.

**UI link** is [here](https://d2l4tym4l7m8ka.cloudfront.net/)
**FE PR link** is [here](https://github.com/nadktk/shop-react-redux-cloudfront/pull/3)

Endpoints:
[get products list](https://elbwtkyhc4.execute-api.eu-west-1.amazonaws.com/dev/products
)
[get product by id](https://elbwtkyhc4.execute-api.eu-west-1.amazonaws.com/dev/products/19e0846b-1d44-4c4e-9e6b-f0bb5f33c3f7
)
create product:
curl --location --request POST 'https://elbwtkyhc4.execute-api.eu-west-1.amazonaws.com/dev/products' \
--header 'Content-Type: application/json' \
--data-raw '{{bodyHere}}'

### Additional (optional) tasks

- [x] (+1) POST /products lambda functions returns error 400 status code if product data is invalid
- [x] (+1) All lambdas return error 500 status code on any error (DB connection, any unhandled error in code)
- [x] (+1) All lambdas do console.log for each incoming requests and their arguments
- [x] (+1) Transaction based creation of product (in case stock creation is failed then related to this stock product is not created and not ready to be used by the end user and vice versa) (https://devcenter.kinvey.com/nodejs/tutorials/bl-transactional-support)

### CODE REVIEW

**Comment by cetrix13** [comment]
> Dear reviewers, I kindly ask you to give me a couple of days to finish the task

Sure, you have two more days until 28.07.2022. I am going to check it then.

File: product-service/services/products.js [hunk 1]
```diff
@@ -1,12 +1,35 @@
...
+import { client } from '../db/client';
 
 class ProductsService {
     async getList() {
-        return Promise.resolve(productData);
+        try {
+            await client.connect();
+                   
+            const products = await client.query(`
+                SELECT * FROM products;
```
Comment (cetrix13) [inline product-service/services/products.js]: You have to run the query with a JOIN on stocks table to get count as well.

File: product-service/services/products.js [hunk 2]
```diff
@@ -1,12 +1,72 @@
...
+
+            return {
+                ...product.rows[0],
+                count: stock.rows[0].count,
+            };
+        } catch (e) {
+            console.log(e);
+
+            throw e;
+        }
```
Comment (cetrix13) [inline product-service/services/products.js:81]: You have to close a connection after all queries are done.

File: product-service/services/products.js [hunk 3]
```diff
@@ -1,12 +1,72 @@
...
+
+            return Promise.resolve(products.rows[0]);
+        } catch (e) {
+            console.log(e);
+
+            throw e;
+        }
+    }
+
+    async create({ count, price, title, description }) {
```
Comment (cetrix13) [inline product-service/services/products.js:56]: I see errors in Postman when I call this [method](https://elbwtkyhc4.execute-api.eu-west-1.amazonaws.com/dev/products). 
image

Though validation of body in the request is correct.

Comment (nadktk) [reply product-service/services/products.js:56]: It's because PR is still in draft and work is in progress. Thank you for the patience

**Pull request merged**